modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStore.java

```diff
@@ -182,23 +198,36 @@ public boolean read(long pageId, ByteBuffer pageBuf, boolean keepCrc) throws Ign
      * @param pageBuf Page buffer to read into.
      * @param checkCrc Check CRC on page.
      * @param keepCrc By default reading zeroes CRC which was on file, but you can keep it in pageBuf if set keepCrc.
+     * @param checkPageOffsetLogically Check page offset by {@link #allocatedBytes} or {@link #size}.
      * @return {@code true} if page has been read successfully, {@code false} if page hasn't been written yet.
      * @throws IgniteInternalCheckedException If reading failed (IO error occurred).
      */
-    private boolean read(long pageId, ByteBuffer pageBuf, boolean checkCrc, boolean keepCrc) throws IgniteInternalCheckedException {
+    private boolean read0(
+            long pageId,
+            ByteBuffer pageBuf,
+            boolean checkCrc,
+            boolean keepCrc,
+            boolean checkPageOffsetLogically
+    ) throws IgniteInternalCheckedException {
         ensure();
 
         try {
-            long off = pageOffset(pageId);
-
             assert pageBuf.capacity() == pageSize : pageBuf.capacity();
             assert pageBuf.remaining() == pageSize : pageBuf.remaining();
             assert pageBuf.position() == 0 : pageBuf.position();
             assert pageBuf.order() == nativeOrder() : pageBuf.order();
-            assert off <= allocatedBytes() : "calculatedOffset=" + off
-                    + ", allocated=" + allocatedBytes() + ", headerSize=" + headerSize + ", filePath=" + filePath;
 
-            int n = readWithFailover(pageBuf, off);
+            long pageOff = pageOffset(pageId);
+
+            if (checkPageOffsetLogically) {
+                assert pageOff <= allocatedBytes() : "calculatedOffset=" + pageOff
```

Comment:
Technically speaking, we could add "pageSize" to "pageOff" in both assertions, because "pageOff" points to a beginning of the page, right?

Reply:
**fileOff** sounds better?

Reply:
This wasn't about the naming, it was about the fact that the assertion is not strict enough, at least if looks like it is

Reply:
I analyzed the code and realized that the check should be left as is, we should not catch an error if the page has not yet been written.

Reply:
Ok, but why? Do we have a code that relies on reading empty pages?